File: DeXShareTarget/Services/GoogleOAuth.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace DeXShareTarget.Services
{
    /// <summary>
    /// PC-side Google Sign-In: OAuth 2.0 Authorization Code + PKCE with a loopback
    /// redirect (http://127.0.0.1:53325). The verified email becomes this PC's identity,
    /// so every same-email device is auto-trusted.
    ///
    /// Credentials live in oauth.local.json (git-ignored — never commit it).
    /// A raw TcpListener is used instead of HttpListener so no URL ACL is required.
    /// </summary>
    public static class GoogleOAuth
    {
        private const int LoopbackPort = 53325;
        private const string RedirectUri = "http://127.0.0.1:53325/";
        private const string AuthorizeUrl = "https://accounts.google.com/o/oauth2/v2/auth";
        private const string TokenUrl = "https://oauth2.googleapis.com/token";

    public record GoogleProfile(string Email, string Name, string Picture, string Sub);

    private static (string ClientId, string ClientSecret)? _cached;

    public static bool IsConfigured() => LoadCredentials() != null;

    /// <summary>Runs the full sign-in flow. Returns the verified profile, or null on failure/cancel.</summary>
    public static async Task<GoogleProfile?> SignInAsync(CancellationToken ct = default)
    {
        var creds = LoadCredentials();
        if (creds == null) return null;

        var verifier = Base64Url(RandomNumberGenerator.GetBytes(32));
        var challenge = Base64Url(SHA256.HashData(Encoding.UTF8.GetBytes(verifier)));
        var state = Guid.NewGuid().ToString("N");

        var authUrl = $"{AuthorizeUrl}?client_id={Uri.EscapeDataString(creds.Value.ClientId)}" +
            $"&redirect_uri={Uri.EscapeDataString(RedirectUri)}&response_type=code" +
            $"&scope={Uri.EscapeDataString("openid email profile")}" +
            $"&code_challenge={challenge}&code_challenge_method=S256&state={state}" +
            $"&prompt=select_account";

        // Open the system browser, then await the loopback redirect
        var codeTask = WaitForCodeAsync(state, ct);
        try
        {
            Process.Start(new ProcessStartInfo(authUrl) { UseShellExecute = true });
        }
        catch { /* browser may fail to launch; the listener still times out */ }

        var code = await codeTask;
        if (code == null) return null;

        // Exchange the authorization code for tokens (desktop-app pattern: secret included)
        using var http = new HttpClient { Timeout = TimeSpan.FromSeconds(15) };
        var form = new FormUrlEncodedContent(new Dictionary<string, string>
        {
            ["code"] = code,
            ["client_id"] = creds.Value.ClientId,
            ["client_secret"] = creds.Value.ClientSecret,
            ["redirect_uri"] = RedirectUri,
            ["grant_type"] = "authorization_code",
            ["code_verifier"] = verifier
        });
        var response = await http.PostAsync(TokenUrl, form, ct);
        if (!response.IsSuccessStatusCode) return null;

        using var json = JsonDocument.Parse(await response.Content.ReadAsStringAsync(ct));
        var idToken = json.RootElement.TryGetProperty("id_token", out var t) ? t.GetString() : null;
        if (string.IsNullOrEmpty(idToken)) return null;

        // id_token is a JWT: header.payload.signature — the profile lives in the payload
        var parts = idToken.Split('.');
        if (parts.Length != 3) return null;
        try
        {
            var decoded = Encoding.UTF8.GetString(Convert.FromBase64String(Base64Pad(parts[1])));
            using var claims = JsonDocument.Parse(decoded);
            var email = claims.RootElement.TryGetProperty("email", out var e) ? e.GetString() : null;
            if (string.IsNullOrEmpty(email)) return null;

            var profile = new GoogleProfile(
                Email: email,
                Name: claims.RootElement.TryGetProperty("name", out var n) ? n.GetString() ?? "" : "",
                Picture: claims.RootElement.TryGetProperty("picture", out var p) ? p.GetString() ?? "" : "",
                Sub: claims.RootElement.TryGetProperty("sub", out var s) ? s.GetString() ?? "" : ""
            );
            SaveProfile(profile);
            return profile;
        }
        catch { return null; }
    }

    /// <summary>Persists the signed-in profile so the UI can show it after restarts.</summary>
    public static void SaveProfile(GoogleProfile profile)
    {
        try
        {
            var dir = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "DeX");
            Directory.CreateDirectory(dir);
            File.WriteAllText(Path.Combine(dir, "google_profile.json"), JsonSerializer.Serialize(profile));
        }
        catch { }
    }

    /// <summary>The last signed-in profile, or null when never signed in.</summary>
    public static GoogleProfile? LoadProfile()
    {
        try
        {
            var path = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "DeX", "google_profile.json");
            if (!File.Exists(path)) return null;
            return JsonSerializer.Deserialize<GoogleProfile>(File.ReadAllText(path));
        }
        catch { return null; }
    }

    /// <summary>Removes the stored profile (sign-out).</summary>
    public static void SignOut()
    {
        try
        {
            var path = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "DeX", "google_profile.json");
            if (File.Exists(path)) File.Delete(path);
        }
        catch { }
    }

        private static async Task<string?> WaitForCodeAsync(string expectedState, CancellationToken ct)
        {
            using var listener = new TcpListener(IPAddress.Loopback, LoopbackPort);
            listener.Start();
            try
            {
                using var timeout = CancellationTokenSource.CreateLinkedTokenSource(ct);
                timeout.CancelAfter(TimeSpan.FromMinutes(2));
                while (!timeout.IsCancellationRequested)
                {
                    var client = await listener.AcceptTcpClientAsync(timeout.Token);
                    try
                    {
                        using var stream = client.GetStream();
                        var requestBytes = new byte[8192];
                        var read = await stream.ReadAsync(requestBytes.AsMemory(0, requestBytes.Length), timeout.Token);
                        var request = Encoding.ASCII.GetString(requestBytes, 0, read);
                        var firstLine = request.Split('\r', '\n')[0];
                        var path = firstLine.Split(' ').Length > 1 ? firstLine.Split(' ')[1] : "/";
                        var query = path.Contains('?') ? path[(path.IndexOf('?') + 1)..] : "";
                        var code = GetQueryValue(query, "code");
                        var state = GetQueryValue(query, "state");
                        var success = code != null && state == expectedState;

                        var body = success
                            ? "<html><body style=\"font-family:sans-serif;text-align:center;margin-top:3em\"><h3>DeX signed in — you can close this tab.</h3></body></html>"
                            : "<html><body style=\"font-family:sans-serif;text-align:center;margin-top:3em\"><h3>Sign-in failed or cancelled.</h3></body></html>";
                        var responseText = $"HTTP/1.1 200 OK\r\nContent-Type: text/html; charset=utf-8\r\nContent-Length: {Encoding.UTF8.GetByteCount(body)}\r\nConnection: close\r\n\r\n{body}";
                        var responseBytes = Encoding.UTF8.GetBytes(responseText);
                        await stream.WriteAsync(responseBytes.AsMemory(0, responseBytes.Length), timeout.Token);

                        if (success) return code;
                    }
                    catch { /* keep waiting for the right redirect */ }
                }
            }
            catch { }
            return null;
        }

        private static string? GetQueryValue(string query, string key)
        {
            foreach (var pair in query.Split('&'))
            {
                var eq = pair.IndexOf('=');
                if (eq <= 0) continue;
                var k = Uri.UnescapeDataString(pair[..eq]);
                if (k == key) return Uri.UnescapeDataString(pair[(eq + 1)..]);
            }
            return null;
        }

        private static string Base64Url(byte[] bytes) =>
            Convert.ToBase64String(bytes).TrimEnd('=').Replace('+', '-').Replace('/', '_');

        private static string Base64Pad(string s)
        {
            switch (s.Length % 4)
            {
                case 2: return s + "==";
                case 3: return s + "=";
                default: return s;
            }
        }

        private static (string ClientId, string ClientSecret)? LoadCredentials()
        {
            if (_cached != null) return _cached;
            var file = FindCredentialsFile();
            if (file == null) return null;
            try
            {
                using var doc = JsonDocument.Parse(File.ReadAllText(file));
                var id = doc.RootElement.TryGetProperty("desktopClientId", out var i) ? i.GetString() : null;
                var secret = doc.RootElement.TryGetProperty("desktopClientSecret", out var s) ? s.GetString() : null;
                if (!string.IsNullOrEmpty(id) && !string.IsNullOrEmpty(secret)) _cached = (id, secret);
            }
            catch { }
            return _cached;
        }

        private static string? FindCredentialsFile()
        {
            var candidates = new List<string>
            {
                Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "oauth.local.json"),
                Path.Combine(Directory.GetCurrentDirectory(), "oauth.local.json")
            };
            // Walk up from the output directory toward the source root (dev runs)
            var dir = new DirectoryInfo(AppDomain.CurrentDomain.BaseDirectory);
            for (var i = 0; i < 6 && dir != null; i++, dir = dir.Parent)
            {
                candidates.Add(Path.Combine(dir.FullName, "oauth.local.json"));
                candidates.Add(Path.Combine(dir.FullName, "DeXShareTarget", "oauth.local.json"));
            }
            return candidates.FirstOrDefault(File.Exists);
        }
    }
}

File: core/network/src/desktopMain/kotlin/com/dexstudios/dex/core/network/auth/GoogleOAuth.kt
package com.dexstudios.dex.core.network.auth

import co.touchlab.kermit.Logger
import com.dexstudios.dex.core.network.DeXPorts
import io.ktor.client.HttpClient
import io.ktor.client.request.forms.FormDataContent
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.client.statement.bodyAsText
import io.ktor.http.Parameters
import io.ktor.http.isSuccess
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.withTimeoutOrNull
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive
import java.awt.Desktop
import java.io.File
import java.net.URI
import java.net.URLEncoder
import java.security.MessageDigest
import java.security.SecureRandom
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.Base64
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

object GoogleOAuth {

    // Served by DeXServer's dedicated loopback listener (legacy WPF Kestrel parity — this
    // exact URI is registered in the Google Cloud Console client, never change it casually).
    private const val REDIRECT_URI = "http://127.0.0.1:${DeXPorts.OAUTH_CALLBACK}/local/oauth/callback"
    private const val AUTHORIZE_URL = "https://accounts.google.com/o/oauth2/v2/auth"
    private const val TOKEN_URL = "https://oauth2.googleapis.com/token"

    private val pending = ConcurrentHashMap<String, CompletableDeferred<String?>>()

    val baseDirectory: File by lazy {
        val osName = System.getProperty("os.name").lowercase()
        if (osName.contains("win")) {
            val localAppData = System.getenv("LOCALAPPDATA") ?: System.getProperty("user.home")
            File(localAppData, "DeX")
        } else {
            File(System.getProperty("user.home"), ".dex_settings")
        }
    }

    @Serializable
    data class GoogleProfile(val email: String, val name: String, val picture: String, val sub: String)

    private var cachedCredentials: Pair<String, String>? = null

    private val json = Json { ignoreUnknownKeys = true }

    fun isConfigured(): Boolean = loadCredentials() != null

    class CallbackResult(val success: Boolean, val error: String?)

    fun handleCallback(state: String, code: String?): Pair<Boolean, String?> {
        val tcs = pending.remove(state)
        if (tcs != null) {
            if (code.isNullOrEmpty()) {
                tcs.complete(null)
                return Pair(false, "No authorization code in redirect.")
            }
            tcs.complete(code)
            return Pair(true, null)
        }
        return Pair(false, "No pending sign-in for this state (may have expired).")
    }

    suspend fun signInAsync(): GoogleProfile? {
        val creds = loadCredentials() ?: return null

        val secureRandom = SecureRandom()
        val verifierBytes = ByteArray(32)
        secureRandom.nextBytes(verifierBytes)
        val verifier = base64Url(verifierBytes)

        val digest = MessageDigest.getInstance("SHA-256")
        val challengeBytes = digest.digest(verifier.toByteArray(Charsets.UTF_8))
        val challenge = base64Url(challengeBytes)

        val state = UUID.randomUUID().toString().replace("-", "")

        val authUrl = buildString {
            append(AUTHORIZE_URL)
            append("?client_id=").append(URLEncoder.encode(creds.first, "UTF-8"))
            append("&redirect_uri=").append(URLEncoder.encode(REDIRECT_URI, "UTF-8"))
            append("&response_type=code")
            append("&scope=").append(URLEncoder.encode("openid email profile", "UTF-8"))
            append("&code_challenge=").append(challenge)
            append("&code_challenge_method=S256")
            append("&state=").append(state)
            append("&prompt=select_account")
        }

        val deferred = CompletableDeferred<String?>()
        pending[state] = deferred

        try {
            if (Desktop.isDesktopSupported() && Desktop.getDesktop().isSupported(Desktop.Action.BROWSE)) {
                Desktop.getDesktop().browse(URI(authUrl))
            } else {
                pending.remove(state)
                log("Desktop browsing not supported.")
                return null
            }
        } catch (e: Exception) {
            pending.remove(state)
            log("Browser failed to launch: ${e.message}")
            return null
        }

        val code = withTimeoutOrNull(230_000L) {
            deferred.await()
        }

        if (code == null) {
            pending.remove(state)?.complete(null)
            return null
        }

        HttpClient().use { http ->
            val formParameters = Parameters.build {
                append("code", code)
                append("client_id", creds.first)
                append("client_secret", creds.second)
                append("redirect_uri", REDIRECT_URI)
                append("grant_type", "authorization_code")
                append("code_verifier", verifier)
            }

            val response: HttpResponse = http.post(TOKEN_URL) {
                setBody(FormDataContent(formParameters))
            }

            if (!response.status.isSuccess()) return null

            val responseBody = response.bodyAsText()
            val jsonResponse = json.decodeFromString<JsonObject>(responseBody)
            val idToken = jsonResponse["id_token"]?.jsonPrimitive?.contentOrNull ?: return null

            val parts = idToken.split(".")
            if (parts.size != 3) return null

            try {
                val decoded = String(Base64.getUrlDecoder().decode(base64Pad(parts[1])), Charsets.UTF_8)
                val claims = json.decodeFromString<JsonObject>(decoded)

                val email = claims["email"]?.jsonPrimitive?.content ?: return null
                val profile = GoogleProfile(
                    email = email,
                    name = claims["name"]?.jsonPrimitive?.content ?: "",
                    picture = claims["picture"]?.jsonPrimitive?.content ?: "",
                    sub = claims["sub"]?.jsonPrimitive?.content ?: "",
                )
                saveProfile(profile)
                return profile
            } catch (e: Exception) {
                return null
            }
        }
    }

    fun saveProfile(profile: GoogleProfile) {
        try {
            if (!baseDirectory.exists()) {
                baseDirectory.mkdirs()
            }
            val file = File(baseDirectory, "google_profile.json")
            file.writeText(json.encodeToString(profile))
        } catch (e: Exception) {
            Logger.i("GoogleOAuth error: ${e.message}")
        }
    }

    fun loadProfile(): GoogleProfile? {
        try {
            val file = File(baseDirectory, "google_profile.json")
            if (!file.exists()) return null
            return json.decodeFromString<GoogleProfile>(file.readText())
        } catch (e: Exception) {
            return null
        }
    }

    fun signOut() {
        try {
            val file = File(baseDirectory, "google_profile.json")
            if (file.exists()) file.delete()
        } catch (e: Exception) {
            Logger.i("GoogleOAuth error: ${e.message}")
        }
    }

    private fun log(message: String) {
        try {
            if (!baseDirectory.exists()) {
                baseDirectory.mkdirs()
            }
            val file = File(baseDirectory, "oauth.log")
            val timestamp = LocalDateTime.now().format(DateTimeFormatter.ISO_DATE_TIME)
            file.appendText("[$timestamp] $message\n")
        } catch (e: Exception) {
            Logger.i("GoogleOAuth file I/O error: ${e.message}")
        }
    }

    private fun base64Url(bytes: ByteArray): String = Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)

    private fun base64Pad(s: String): String = when (s.length % 4) {
        2 -> "$s=="
        3 -> "$s="
        else -> s
    }

    private fun loadCredentials(): Pair<String, String>? {
        if (cachedCredentials != null) return cachedCredentials
        val file = findCredentialsFile() ?: return null
        try {
            val content = file.readText()
            val doc = json.decodeFromString<JsonObject>(content)
            val id = doc["desktopClientId"]?.jsonPrimitive?.content
            val secret = doc["desktopClientSecret"]?.jsonPrimitive?.content
            if (!id.isNullOrEmpty() && !secret.isNullOrEmpty()) {
                cachedCredentials = Pair(id, secret)
            }
        } catch (e: Exception) {
            Logger.i("GoogleOAuth file I/O error: ${e.message}")
        }
        return cachedCredentials
    }

    private fun findCredentialsFile(): File? {
        val currentDir = File(System.getProperty("user.dir"))
        val candidates = mutableListOf(
            File(currentDir, "oauth.local.json"),
        )

        var dir: File? = currentDir
        for (i in 0..5) {
            if (dir == null) break
            candidates.add(File(dir, "oauth.local.json"))
            candidates.add(File(dir, "DeXShareTarget/oauth.local.json"))
            dir = dir.parentFile
        }

        return candidates.firstOrNull { it.exists() && it.isFile }
    }
}
